068-access-policies-in-jwt.md

```diff
@@ -0,0 +1,316 @@
+As the documentation states:
+
+> In addition to the Kafka custom resource configuration, the JAR file containing the custom authorizer class along with its dependencies must be available on the classpath of the Kafka broker.
+
+This proposal aims to add the JWT-based authorizer to Strimzi, so it can be selected in the configuration as
+another option apart from the simple and Keycloak ones, without having to implement a custom authorizer.
+
+## Motivation
+
+Having a generic solution is always preferable as it allows for less effort to be maintained.
+While we cannot achieve a perfect generic solution due to the nature of OAuth2, and how different
+identity providers are implemented, we can strive to create a solution as generic as possible.
+
+Passing additional data in the JWT seems to be supported by the majority of identity providers.
+Therefore, we can use this to pass the access policies in the JWT.
+This solution will work in most cases and with most identity providers, however it will not
+scale to infinity.
+It will, however, be decoupled from Keycloak and therefore be more generic.
+
+## Why JWT as an authorization mechanism?
+
+The official [JWT](https://jwt.io/introduction) lists some use cases:
+
+> Authorization: This is the most common scenario for using JWT. Once the user is logged in, each subsequent request will include the JWT, allowing the user to access routes, services, and resources that are permitted with that token. Single Sign On is a feature that widely uses JWT nowadays, because of its small overhead and its ability to be easily used across different domains.
+
+JWT supports custom claims, which can be used to pass additional claims/roles.
+For example [Spring Security](https://docs.spring.io/spring-authorization-server/reference/guides/how-to-custom-claims-authorities.html) can pass both claims and roles in the JWT,
+which can be used to authorize the user.
+
+All OAuth2 providers support custom claims in the JWT in one way or another.
+We have listed the most popular ones in the [Support for roles or claims in popular identity providers](#support-for-roles-or-claims-in-popular-identity-providers) section with links to the documentation
+on how custom claims or roles are supported.
+
+The concept of roles can be directly mapped to the concept of access policies in Kafka.
+
+## Proposal
+
+Identity providers support passing custom fields in the JWT token.
+Some providers allow the data type to be string, while others allow a list of strings.
+There are not many (if any) identity providers that allow a list of objects.
+
+Therefore, in this proposal we allow the ACLs to be passed as either.
+e.g. for a list of strings:
+```json
+{
+  "acls": [
+    "cluster_x:topic_1:read",
+    ":topic2:write"
+  ]
+}
+```
+
+and for a string:
+```json
+{
+  "acls": "cluster_x:topic_1:read,topic2:write"
+}
+```
+
+The field name is configurable with default value `acls`.
+
+### Using roles to pass ACLs
```

Comment:
@tombentley Adding a section to cover how roles are mapped to kafka ACLs. I am using roles as these are supported by all IDPs. If an IDP supports more advanced mechanisms of authorization or JWT manipulation, then it could also be used.

Reply:
I think the idea is to piggyback on the roles by naming them using this custom ACL syntax. However, that may look quite hackish as the UI used to configure that may offer a drop down list of all the different ACLs and it may be quite confusing.

Reply:
I can see why you would see this as hack-ish. Some IDPs support other ways of enriching the JWT, but they are specific to the IDP.  I chose to give the example with roles as it is what most people would know and what would work under any IDP (albeit with the drawbacks that you mention). For the same reason, I'd give the same example in the documentation.
I have used fine-grained roles like this before and can say it works fine for people.
For the UI concerns, I can talk to a team member who has more experience with this and come back to you.